File: code/cpp/ch09/practice/warmup_04_last_occurrence.cpp
/*
 * Warmup 4: Last Occurrence
 * ===========================
 * Chapter 9: Finding Needles -- The Power of Searching
 *
 * PROBLEM:
 *   Given a SORTED array and a target, find the index of the LAST
 *   (rightmost) occurrence of target. Return -1 if not found.
 *
 * EXAMPLES:
 *   solve({1,2,2,2,3,4}, 2) -> 3
 *   solve({1,1,1,1,1}, 1)   -> 4
 *   solve({1,3,5,7}, 5)     -> 2
 *   solve({1,3,5,7}, 4)     -> -1
 *
 * CONSTRAINTS:
 *   0 <= arr.size() <= 10^5
 *   Array is sorted in non-decreasing order
 *
 * INSTRUCTIONS:
 *   Replace the body with your solution.
 *   Do NOT use std::upper_bound.
 */

#include <iostream>
#include <vector>
using namespace std;

int solve(vector<int> arr, int target) {
    // TODO: Replace this with your solution
    return -1;
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) cin >> arr[i];
    int target;
    cin >> target;
    cout << solve(arr, target) << endl;
    return 0;
}

File: code/python/ch09/practice/warmup_04_last_occurrence.py
"""
Warmup 4: Last Occurrence
==============================
Chapter 9: Finding Needles — The Power of Searching

PROBLEM
-------
Given a sorted array of integers (which may contain duplicates) and a
target value, find the index of the LAST occurrence of the target.
If the target is not present, return -1.

INPUT FORMAT
------------
First line: space-separated integers (a sorted array).
Second line: a single integer (the target).

OUTPUT FORMAT
-------------
A single integer: the index of the last occurrence, or -1 if not found.

CONSTRAINTS
-----------
- 0 <= len(arr) <= 10^5
- -10^9 <= arr[i] <= 10^9
- The array is sorted in non-decreasing order.

EXAMPLES
--------
Input:
  1 2 2 2 3 4
  2
Output: 3

Input:
  1 1 1 1 1
  1
Output: 4

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(arr: list[int], target: int) -> int:
    """Return the index of the last occurrence of target, or -1."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    data = list(map(int, input().split()))
    target = int(input())
    print(solve(data, target))
